### PR TITLE
pageserver: basic cancel/timeout for remote storage operations

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

Various places in remote storage were not subject to a timeout (thereby stuck TCP connections could hold things up), and did not respect a cancellation token (so things like timeline deletion or tenant detach would have to wait arbitrarily long).

## Summary of changes

- Add download_cancellable and upload_cancellable helpers, and use them in all the places we wait for remote storage operations (with the exception of initdb downloads, where it would not have been safe).
- Add a cancellation token arg to `download_retry`.
- Use cancellation token args in various places that were missing one per #5066 

Closes: #5066 

Why is this only "basic" handling?
- Doesn't express difference between shutdown and errors in return types, to avoid refactoring all the places that use an anyhow::Error (these should all eventually return a more structured error type)
- Implements timeouts on top of remote storage, rather than within it: this means that operations hitting their timeout will lose their semaphore permit and thereby go to the back of the queue for their retry.
- Doing a nicer job is tracked in https://github.com/neondatabase/neon/issues/6096

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant/delete.rs [hunk 1]
```diff
@@ -322,9 +322,15 @@ impl DeleteTenantFlow {
...
         // Detach currently uses remove_dir_all so in case of a crash we can end up in a weird state.
         if let Some(remote_storage) = &remote_storage {
-            create_remote_delete_mark(conf, remote_storage, &tenant.tenant_shard_id)
-                .await
-                .context("remote_mark")?
+            create_remote_delete_mark(
+                conf,
+                remote_storage,
+                &tenant.tenant_shard_id,
+                // Can't use tenant.cancel, it's already shut down.  TODO: wire in an appropriate token
```
Comment (arpad-m) [inline pageserver/src/tenant/delete.rs:329]: which token would be appropriate here? the connection level token? or the ps shutdown token?

Comment (jcsp) [reply pageserver/src/tenant/delete.rs:329]: The process-wide one -- i have a followup here that adds a cancellation token on the top level of TenantManager, and does the refactoring required to move the tenant deletion flow into TenantManager to use it https://github.com/neondatabase/neon/pull/6105

**Review by koivunej — COMMENTED** [review]
I don't think this should be approved because of work required to now undo this PR. If anything, merge only right before next release.

**Comment by jcsp** [comment]
This PR is +166 lines, -45 lines.  It will not be exceptionally onerous to unpick when implementing cancellation in the remote_storage layer later: it'll basically be a case of replacing all uses of upload_cancellable and download_cancellable.

**Review by koivunej — APPROVED** [review]
I will undo these in the other PR once I get there.

**Review by arpad-m — APPROVED** [review]

**Pull request merged**